ir/json_generator.h

```diff
@@ -55,156 +56,206 @@ class JSONGenerator {
-        void operator()(const T &value) const {
-            out << R"("value" : )";
-            generator.generate(value);
+    void end_vector(output_state_t prev) {
+        --indent;
+        if (output_state == VEC_MID)
+            out << std::endl << indent;
+        else if (output_state != VEC_START)
+            BUG("invalid json output state in end_vector");
+        out << ']';
+        if ((output_state = prev) == OBJ_AFTERTAG) output_state = OBJ_MID;
+    }
+
+    output_state_t begin_object() {
+        output_state_t rv = output_state;
+        output_state = OBJ_START;
+        return rv;
+    }
+
+    void end_object(output_state_t prev) {
+        switch (output_state) {
+            case OBJ_START:
+                out << "{}";
+                break;
+            case OBJ_MID:
+                out << std::endl << --indent << '}';
+                break;
+            case OBJ_END:
+                break;
+            default:
+                BUG("invalid json output state in end_object");
+                break;
         }
-    };
+        if ((output_state = prev) == OBJ_AFTERTAG) output_state = OBJ_MID;
+    }
 
- public:
-    indent_t indent;
+    template <typename T>
+    void emit(const T &val) {
+        switch (output_state) {
+            case VEC_MID:
+                out << ',';
+                /* fall through */
+            case VEC_START:
+                out << std::endl << indent;
+                output_state = VEC_MID;
+                break;
+            case OBJ_AFTERTAG:
+                output_state = OBJ_MID;
+                break;
+            case TOP:
+                break;
+            case OBJ_START:
+                output_state = OBJ_END;
+                break;
+            case OBJ_MID:
+            case OBJ_END:
+                BUG("invalid json output state for emit(obj)");
+        }
+        generate(val);
+        if (output_state == TOP) out << std::endl;
+    }
 
-    explicit JSONGenerator(std::ostream &out, bool dumpSourceInfo = false)
-        : out(out), dumpSourceInfo(dumpSourceInfo) {}
+    void emit_tag(const char *tag) {
+        switch (output_state) {
+            case OBJ_START:
+                out << '{' << std::endl << ++indent;
+                break;
+            case OBJ_MID:
+                out << ',' << std::endl << indent;
+                break;
+            case TOP:
+            case VEC_START:
+            case VEC_MID:
+            case OBJ_AFTERTAG:
+            case OBJ_END:
+                BUG("invalid json output state for emit_tag");
+        }
+        out << '\"' << cstring(tag).escapeJson() << "\" : ";
+        output_state = OBJ_AFTERTAG;
+    }
+
+    template <typename T>
+    void emit(const char *tag, const T &val) {
+        emit_tag(tag);
+        output_state = OBJ_MID;
+        generate(val);
+    }
+
+    void emit_tag(cstring tag) { emit_tag(tag.c_str()); }
+    void emit_tag(std::string tag) { emit_tag(tag.c_str()); }
```

Comment:
Do you really want to copy string every time here? Same for `emit<T>(std::string, const T &val)` below.

Reply:
Actually, it probably makes more sense to mostly convert this code to taking `std::string_view` (and allow cstring to be converted to that) so we only need one version of each of these things.

Reply:
Yes, this is very true. Though we are not yet ready for this IIRC. I think we are still relying on implicit conversion from `std::string` to `cstring` and this might make this ambiguous.